Revise “Kubernetes API Concepts” (#25299)
kubernetes/website: content/en/docs/reference/using-api/api-concepts.md

resources to distinguish from retrieving a single resource which is usually called
a **get**. If you sent an HTTP GET request with the `?watch` query parameter,
Kubernetes calls this a **watch** and not a **get** (see
[Efficient detection of changes](#efficient-detection-of-changes) for more details).

For PUT requests, Kubernetes internally classifies these as either **create** or **update**
based on the state of the existing object. An **update** is different from a **patch**; the
HTTP verb for a **patch** is PATCH.

{{< note >}}
The *server-side apply* concept differentiates between `Apply` and `Update` operations. That
`Update` operation is a separate concept from the logical **update** verb described here.

See [Apply and Update](/docs/reference/using-api/server-side-apply/#apply-and-update) for
more details.
{{< /note >}}

## Resource URIs

All resource types are either scoped by the cluster (`/apis/GROUP/VERSION/*`) or to a namespace (`/apis/GROUP/VERSION/namespaces/NAMESPACE/*`). A namespace-scoped resource type will be deleted when its namespace is deleted and access to that resource type is controlled by authorization checks on the namespace scope.

> You can also access collections of resources (for example: listing all Nodes)

Reviewer: ```suggestion
You can also access collections of resources (for example: listing all Nodes).
```

Kubernetes also provides consistent list operations so that API clients can
effectively cache, track, and synchronize the state of resources.

You can view the [API reference](/docs/reference/kubernetes-api/) online,
or read on to learn about the API in general.

<!-- body -->
## Kubernetes API terminology {#standard-api-terminology}

Kubernetes generally leverages standard RESTful terminology to describe the
API concepts:

* A *resource type* is the name used in the URL (`pods`, `namespaces`, `services`)
* All resource types have a concrete representation in JSON (their object schema) which is called a *kind*
* A list of instances of a resource type is known as a *collection*
* A single instance of the resource type is called a *resource*
* Some resource types have one or more *sub-resources*, represented as URI paths below the resource

Most Kubernetes API resource types are
[objects](/docs/concepts/overview/working-with-objects/kubernetes-objects/#kubernetes-objects):
they represent a concrete instance of a concept on the cluster, like a
pod or namespace. A smaller number of API resource types are *virtual* in

> they often represent operations on objects, rather than objects, such as a

Reviewer: nit:  A smaller number of API resource types are *virtual* in that they often represent operations on objects, rather than objects ...

permission check (use a POST with a JSON-encoded body of `SubjectAccessReview`
to the `subjectaccessreviews` resource), or the `eviction` sub-resource of a
Pod (used to trigger
[API-initiated eviction](/docs/concepts/scheduling-eviction/api-eviction/)).

> All objects you can create via the API have a unique object name to allow idempotent creation and retrieval, except that virtual resource types may not have unique names if they are not retrievable, or do not rely on idempotency.

Reviewer: nit: Could break this sentence up (follow-up PR).
The virtual object description is confusing to me.

Reviewer 2: ```suggestion
All objects you can create via the API have a unique (within their container) object name to allow idempotent creation and retrieval, except that virtual resource types may not have unique names if they are not retrievable, or do not rely on idempotency.
```

---

<!-- overview -->
The Kubernetes API is a resource-based (RESTful) programmatic interface
provided via HTTP. It supports retrieving, creating, updating, and deleting
primary resources via the standard HTTP verbs (POST, PUT, PATCH, DELETE,
GET).

> The API includes additional subresources for many objects that allow

Reviewer 2: ```suggestion
The API includes additional subresources for some resources that allow
```

fine grained authorization (such as a separating viewing details for a Pod from
retrieving its logs), and can accept and serve those resources in different
representations for convenience or efficiency.

Kubernetes supports efficient change notifications on resources via *watches*.
Kubernetes also provides consistent list operations so that API clients can
effectively cache, track, and synchronize the state of resources.

You can view the [API reference](/docs/reference/kubernetes-api/) online,
or read on to learn about the API in general.

<!-- body -->
## Kubernetes API terminology {#standard-api-terminology}

Kubernetes generally leverages standard RESTful terminology to describe the
API concepts:

* A *resource type* is the name used in the URL (`pods`, `namespaces`, `services`)

> * All resource types have a concrete representation in JSON (their object schema) which is called a *kind*

Reviewer 2: Many Kinds also have a proto representation. And all resources can produce a tabular representation on request.

> * A list of instances of a resource type is known as a *collection*

Reviewer 2: ```suggestion
* A list of instances of a resource is known as a *collection*
```

Author: I think the original wording is more accurate. If I have a resource `/api/v1/namespaces/default/services/web` then I don't want to list all the instances of that specific web service (the resource); I want to list all the services in the namespace.

Reviewer 2: "resource" doesn't mean a particular object to me. Fortunately I don't have time for another pass so I'll have to be happy with whatever you come up with :)

Author: We don't have a good term that gives us a way to differentiate “Deployment” (the generic feature of Kubernetes that lets you manage ReplicaSets) from a specific resource (in the HTTP sense of the term) representing a specific Deployment in a named namespace.

> * A single instance of the resource type is called a *resource*

Reviewer 2: ```suggestion
* A single instance of the resource type is called an *object*
```

Reviewer 2: (@deads2k should look at this too, he might disagree with me)

Author: I reworded to explain that most resources are also objects.

Reviewer 2: IMO "resource" actually means what you're calling "resource type". It does not mean a specific object of that type.

Author: If we say we're following REST terminology: https://stackoverflow.com/a/10883810 has a relevant quote.

*However* I have seen a lot of Kubernetes contributors implying that, eg `.../deployments/` is a resource - and I've even done that myself. For this page I wanted to be more precise.

Reviewer 2: Unfortunately (?) in k8s land, "resource" has come to be defined by opposition to "kind", the serialization format.

Author: Can we save that for a follow-up PR? _This_ PR isn't introducing a change here - see https://kubernetes.io/docs/reference/using-api/api-concepts/#standard-api-terminology for the live page.

Reviewer 2: that's fine.

> * Some resource types have one or more *sub-resources*, represented as URI paths below the resource

Reviewer 2: ```suggestion
* Some resources have one or more *sub-resources*, represented as URI paths below the resource
```

Author: I'll reword to capture what I think the original author intended to explain.

(used to trigger
[API-initiated eviction](/docs/concepts/scheduling-eviction/api-eviction/)).

All objects you can create via the API have a unique object name to allow idempotent creation and retrieval, except that virtual resource types may not have unique names if they are not retrievable, or do not rely on idempotency.

### API verbs

Almost all object resource types support the standard HTTP verbs - GET, POST, PUT, PATCH,
and DELETE. Kubernetes also uses its own verbs, which are often written lowercase to distinguish
them from HTTP verbs.

Kubernetes uses the term **list** to describe returning a [collection](#collections) of
resources to distinguish from retrieving a single resource which is usually called
a **get**. If you sent an HTTP GET request with the `?watch` query parameter,
Kubernetes calls this a **watch** and not a **get** (see
[Efficient detection of changes](#efficient-detection-of-changes) for more details).

For PUT requests, Kubernetes internally classifies these as either **create** or **update**
based on the state of the existing object. An **update** is different from a **patch**; the
HTTP verb for a **patch** is PATCH.

{{< note >}}

> The *server-side apply* concept differentiates between `Apply` and `Update` operations. That

Reviewer 2: This is confusing, I'm not sure what it's trying to say. SSA is a special kind of PATCH which can either create or update an object.

Author: Ah. The previous page didn't mention an **apply** verb so I misunderstood how things work.

Author: Is there an exhaustive list of verbs in the Kubernetes API? If that's feasible to document, we could add it (separately) to the website.

Reviewer 2: For k8s specific "verbs" this is only missing APPLY. For general HTTP verbs, we do useful things on at least HEAD, maybe OPTIONS, I can't recall.

Author: The trouble with listing verbs is that RBAC adds extra ones (eg **impersonate**). The reader has to know if the verb is used in an RBAC context or for the general API machinery, aside from authnz. I think I won't try to list them all.

I also won't list all the HTTP verbs as the IETF does a great job on that.

Author: Anyway, I think the original feedback https://github.com/kubernetes/website/pull/25299#discussion_r677690861 was valid and is fixed.

---

<!-- overview -->
The Kubernetes API is a resource-based (RESTful) programmatic interface
provided via HTTP. It supports retrieving, creating, updating, and deleting
primary resources via the standard HTTP verbs (POST, PUT, PATCH, DELETE,
GET).

For some resources, the API includes additional subresources that allow
fine grained authorization (such as a separating viewing details for a Pod from
retrieving its logs), and can accept and serve those resources in different
representations for convenience or efficiency.

Kubernetes supports efficient change notifications on resources via *watches*.
Kubernetes also provides consistent list operations so that API clients can
effectively cache, track, and synchronize the state of resources.

You can view the [API reference](/docs/reference/kubernetes-api/) online,

> or read on to learn about the API in general.

Reviewer: nit: Line 27, You could end the sentence with "online".

<!-- body -->
## Kubernetes API terminology {#standard-api-terminology}

Kubernetes generally leverages common RESTful terminology to describe the
API concepts:

* A *resource type* is the name used in the URL (`pods`, `namespaces`, `services`)
* All resource types have a concrete representation (their object schema) which is called a *kind*
* A list of instances of a resource is known as a *collection*
* A single instance of a resource type is called a *resource*, and also usually represents an *object*
* For some resource types, the API includes one or more *sub-resources*, which are represented as URI paths below the resource

Most Kubernetes API resource types are
[objects](/docs/concepts/overview/working-with-objects/kubernetes-objects/#kubernetes-objects):
they represent a concrete instance of a concept on the cluster, like a

> pod or namespace. A smaller number of API resource types are *virtual* in

Reviewer: nit: Are there two types of resource:
- object
- virtual

Author: That's a question for WG API Expression, I think (I don't know the answer).

that they often represent operations on objects, rather than objects, such
as a permission check
(use a POST with a JSON-encoded body of `SubjectAccessReview` to the
`subjectaccessreviews` resource), or the `eviction` sub-resource of a Pod
(used to trigger
[API-initiated eviction](/docs/concepts/scheduling-eviction/api-eviction/)).

### Object names

All objects you can create via the API have a unique object
{{< glossary_tooltip text="name" term_id="name" >}} to allow idempotent creation and
retrieval, except that virtual resource types may not have unique names if they are
not retrievable, or do not rely on idempotency.
Within a {{< glossary_tooltip text="namespace" term_id="namespace" >}}, only one object
of a given kind can have a given name at a time. However, if you delete the object,
you can make a new object with the same name. Some objects are not namespaced (for
example: Nodes), and so their names must be unique across the whole cluster.

### API verbs

> Almost all object resource types support the standard HTTP verbs - GET, POST, PUT, PATCH,

Reviewer: I have a question about: "Almost all object resource types". Is this the same as:
A Kubernetes resource type that represents an object supports the HTTP verbs: GET, POST, ...

When are the Kubernetes verbs (lowercase) used (add this information)?

Author: > When are the Kubernetes verbs (lowercase) used?

Could you clarify that question? I infer you're asking on behalf of the reader. What is it that they want to know?

and DELETE. Kubernetes also uses its own verbs, which are often written lowercase to distinguish
them from HTTP verbs.

Kubernetes uses the term **list** to describe returning a [collection](#collections) of
resources to distinguish from retrieving a single resource which is usually called
a **get**. If you sent an HTTP GET request with the `?watch` query parameter,
Kubernetes calls this a **watch** and not a **get** (see
[Efficient detection of changes](#efficient-detection-of-changes) for more details).

For PUT requests, Kubernetes internally classifies these as either **create** or **update**
based on the state of the existing object. An **update** is different from a **patch**; the

> HTTP verb for a **patch** is PATCH.

Reviewer: This sentence seems redundant:
the HTTP verb for a **patch** is PATCH.

Author: Maybe:
>  the HTTP verb for a **patch** is PATCH, not PUT.

?

## Resource URIs

All resource types are either scoped by the cluster (`/apis/GROUP/VERSION/*`) or to a

> namespace (`/apis/GROUP/VERSION/namespaces/NAMESPACE/*`). A namespace-scoped resource

Reviewer: Could break up the sentence, "A namespace-scoped resource ..."
The API server (?) deletes the namespace-scoped resource when the namespace is deleted. Access to a resource type is controlled by authorization checks on the namespace scope.

* All resource types have a concrete representation in JSON (their object schema) which is called a **kind**;
* A set of instances of a resource type is known as a **collection**;
* A single instance of the resource type is called an **object**, or sometimes a **resource**.

{{< note >}}
In URLs, Kubernetes types such as `pod` and `configmap` are **lowercase**. This is different from
how Kubernetes conventionally renders those types elsewhere, such as the `.kind` field of an
object, where the type is typically written in **UpperCamelCase**.

For example: `"kind": "ConfigMap"`.
{{< /note >}}

All resource types are either scoped by the cluster (`/apis/<GROUP>/<VERSION>/*`) or to a namespace (`/apis/<GROUP>/<VERSION>/namespaces/<NAMESPACE>/*`). A namespace-scoped resource type will be deleted when its namespace is being deleted, and access to a namespaced resource type is controlled by authorization checks on the namespace scope. The following paths are used to retrieve collections and resources:

* Cluster-scoped resources:

  * `GET /apis/<GROUP>/<VERSION>/<RESOURCETYPE>` - return the collection of resources of the resource type

>   * `GET /apis/<GROUP>/<VERSION>/<RESOURCETYPE>/<NAME` - return the resource with NAME under the resource type

Reviewer 3: It is okay to add additional decorators if that improves the readability. However, as shown by line 42, it is much easier to introduce typos when such decorators are added, i.e. there is a missing `>` at the end.

Author: I've omitted this change from the latest revision.

---

<!-- overview -->

> This page describes Kubernetes' API and conventions for interacting with it.

Reviewer 3: Why this revision?

Author: Concept documentation doesn't belong in the API reference section.
Should we:
 - focus on the reference nature and make the tone of the page match its current home
 - migrate the page into the concepts section and add redirects
?

This change adapts the page to its current home.

Reviewer: - I think it would help to retitle this PR.
- I will read through the changes, but I think the word "concepts" is used generically here (similar to conventions).
You could remove the original sentence and not add a replacement sentence. The following two sentences are long and could possibly be broken up.
- Also, the concepts section has a related page, `concepts/overview/kubernetes-api.md` and then there is the page,
`reference/using-api/_index.md` (`This section provides reference information for the Kubernetes API.`)

Author: > the concepts section has a related page, `concepts/overview/kubernetes-api.md` and then there is the page,
`reference/using-api/_index.md`
> (This section provides reference information for the Kubernetes API.)

I'm fairly convinced that having this information spread across two pages, with the current split, is unhelpful to readers.

Reviewer: OK. The concept section has enough information to introduce the k8s APIs and provide links to more information. The k8s API concepts overview page provides that (https://kubernetes.io/docs/concepts/overview/kubernetes-api/).
One section that could possibly move to the reference section: https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-specification

<!-- body -->
The Kubernetes API is a resource-based (RESTful) programmatic interface provided via HTTP. It supports retrieving, creating,

> updating, and deleting primary resources via the standard HTTP verbs (POST, PUT, PATCH, DELETE, GET). The API includes additional subresources for many objects that allow fine grained authorization (such as binding a pod to a node), and can accept and serve those resources in different representations for convenience or efficiency. Kubernetes also supports efficient change notifications on resources via "watches" and consistent lists, so that API clients can effectively cache, track, and synchronize the state of resources.

Reviewer 3: For such a long line, do you have any suggestions how can I easily find out what have been changed?
Are these changes absolutely necessary?

Author: The original has a long line, and GitHub's diff detection isn't great for documentation.
If you fetch this PR locally you can run eg
`git diff --word-diff=color --word-diff-regex=.` which shows character-by-character changes.

Reviewer: Is replacing "other components" with "API clients" the correct substitution?
I think "other components" refers to components in the cluster that act as "API clients". When I read "API clients" I think of external HTTP clients.

`to allow other cluster components to effectively cache and synchronize the state of resources`.

Author: Maybe “API clients (including other control plane components)”? A watcher can be part of the control plane, or it can be a Pod, or it can be `kubectl`.

Author: I've omitted this change from the latest revision.

> ## Kubernetes API terminology {#standard-api-terminology}

Reviewer 3: Why adding "Kubernetes" here and changing the title?
What else can it be if we omit "Kubernetes"?

Author: The previous title, “Standard API terminology”, implies some common set of principles used across a group of technologies. However, these terms are specific to Kubernetes.

Reviewer: The change implies that Kubernetes does not follow RESTful standards or has a different way of representing resources?
`Kubernetes generally leverages standard RESTful terminology to describe the API concepts:`

Author: ReST is not a _standard_, it's more of a set of principles. Kubernetes does not closely implement those principles as laid out in the ReST paper (and that's fine).

Reviewer 4: From a technical writing standpoint:

- @sftim is correct, REST is a set of principles, not a strictly defined standard. 
- Kubernetes follows RESTful principles as much as any other API I've ever worked with does, which is to say "mostly". 

I don't see any problem with this subtitle and I don't think further discussion should block merging this PR.

> Most Kubernetes API resource types are [objects](/docs/concepts/overview/working-with-objects/kubernetes-objects/#kubernetes-objects): they represent a concrete instance of a concept on the cluster, such as a {{< glossary_tooltip term_id="pod" >}} or {{< glossary_tooltip term_id="namespace" >}}. A smaller number of API resource types are "virtual" - they often represent operations rather than objects, such as a permission check (use a POST with a JSON-encoded body of kind `SubjectAccessReview` to the `subjectaccessreviews` resource). All objects have a unique [names](/docs/concepts/overview/working-with-objects/names/#names) to allow idempotent creation and retrieval, but virtual resource types may not have unique names if they are not retrievable or do not rely on idempotency.

Reviewer 3: Why do we keep adding this glossary_tooltip everywhere?
If a user is already looking into this page, and he/she wants to learn more about the API, how likely he/she doesn't know what is a Pod or what is a namespace?

Author: Everyone has to start somewhere @tengqm. What I _don't_ know is the gaps in every reader's knowledge.

I haven't seen tooltips as unhelpful. Is there a disadvantage?

Reviewer: I need to look at the rendered page to determine if *this* page needs a tooltip about pods and namespaces.
There are plenty of other pages that engage the reader about `Pod` and `Namespace`.

I am more interested in knowing what "virtual" means in this context?

Author: `SubjectAccessReview` looks like good example of a virtual resource, according to the existing definition. This page is the first place I came across that wording.

Author: I've omitted this change from the latest revision.

Kubernetes relies on terminology from [ReST](https://en.wikipedia.org/wiki/Representational_state_transfer)
to name and describe several API concepts:

* A **resource type** is the name used in the URL (for example: `pods`, `namespaces`, `services`);
* All resource types have a concrete representation in JSON (their object schema) which is called a **kind**;
* A set of instances of a resource type is known as a **collection**;
* A single instance of the resource type is called an **object**, or sometimes a **resource**.

{{< note >}}

> In URLs, Kubernetes types such as `pod` and `configmap` are **lowercase**. This is different from

Reviewer 3: This sentence is polluting the who section. If we are talking about "Kubernetes kinds", we should say `Pod` and `ConfigMap`, if we are instead talking about types, we use their lower cased plurals `pods` and `configmaps`.

Author: Let's try to find another way that ensures that readers know why the case in the API is different from the rest of the documentation.

Reviewer: Is it important to mention capitalization conventions in this reference?
I would hope that readers interpret the object specification correctly. I would not question a lowercase term in a URL.

Author: > Is it important to mention capitalization conventions in this reference?

I think so - I'd been using Kubernetes for quite a while before I understood _why_ RBAC uses lowercase even though tools like `kubectl` print similar-looking names in UpperCamelCase. I'd like there to be at least one place in the documentation where someone can find that answered.

Reviewer 3: This revision is wrong. In URLs, we only have lowercase plurals for resources.

> how Kubernetes conventionally renders those types elsewhere, such as the `.kind` field of an

Reviewer 3: The manifests is the exact location where the API specifications are codified. We should not call them "elsewhere". On the contrary, the CLI allows for a more ambiguous spelling for users convenience. Right?

Author: The comparison is between manifests and API URLs.

Reviewer 3: If we really want to clarify the concepts, maybe one of the authoritative source of  this is this: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md

There you can find how the terms are defined:

> * Kind the name of a particular object schema (e.g. the "Cat" and "Dog" kinds would have different attributes and properties)
> * Resource a representation of a system entity, sent or retrieved as JSON via HTTP to the server. Resources are exposed via:
> * Collections - a list of resources of the same type, which may be queryable
> * Elements - an individual resource, addressable via a URL.

> Each resource typically accepts and returns data of a single kind. A kind may be accepted or returned by multiple resources that reflect specific use cases. For instance, the kind "Pod" is exposed as a "pods" resource that allows end users to create, update, and delete pods ...

Author: I've omitted this change from the latest revision.

object, where the type is typically written in **UpperCamelCase**.

For example: `"kind": "ConfigMap"`.
{{< /note >}}

> All resource types are either scoped by the cluster (`/apis/<GROUP>/<VERSION>/*`) or to a namespace (`/apis/<GROUP>/<VERSION>/namespaces/<NAMESPACE>/*`). A namespace-scoped resource type will be deleted when its namespace is being deleted, and access to a namespaced resource type is controlled by authorization checks on the namespace scope. The following paths are used to retrieve collections and resources:

Reviewer: Possible edit:
A namespace-scoped resource is deleted when its namespace is destroyed[removed], and access to a namespace-scoped resource is controlled by authorization checks on the namespace scope. The following paths are used to retrieve collections and resources:

Reviewer 3: > A namespace-scoped resource type will be deleted when its namespace is being deleted

This statement makes no sense. Only resources are deleted, no the resource types.
@kbhawkey's suggestion gets the point.

Author: I've omitted this change from the latest revision.

* Cluster-scoped resources:

  * `GET /apis/<GROUP>/<VERSION>/<RESOURCETYPE>` - return the collection of resources of the resource type
  * `GET /apis/<GROUP>/<VERSION>/<RESOURCETYPE>/<NAME` - return the resource with NAME under the resource type

* Namespace-scoped resources:

  * `GET /apis/<GROUP>/<VERSION>/<RESOURCETYPE>` - return the collection of all instances of the resource type across all namespaces
  * `GET /apis/<GROUP>/<VERSION>/namespaces/<NAMESPACE>/<RESOURCETYPE>` - return collection of all instances of the resource type in NAMESPACE
  * `GET /apis/<GROUP>/<VERSION>/namespaces/<NAMESPACE>/<RESOURCETYPE>/<NAME>` - return the instance of the resource type with NAME in NAMESPACE

Since a namespace is a cluster-scoped resource type, you can retrieve the list of all namespaces with `GET /api/v1/namespaces` and details about a particular namespace with `GET /api/v1/namespaces/<NAME>`.

> Almost all object resource types support the standard HTTP verbs - GET, POST, PUT, PATCH, and DELETE. Kubernetes uses the term **List** to describe returning a collection of resources to distinguish from retrieving a single resource which is usually called a **Get**.

Reviewer 3: Do NOT change `list` to `List` or `get` to `Get` to cause more confusion, please.

Reviewer 3: There are some reference for mapping verbs. Let's be careful when introducing this kind of spelling changes.
https://kubernetes.io/docs/reference/access-authn-authz/authorization/#determine-the-request-verb

Author: I think I'd got confused between `watch` (the verb) and `Watch` (the noun). It's hard to document; I'm learning as I go to be honest.

Author: I've revised my approach here (having learned the API enough to understand what this page is trying to say).

Some resource types have one or more sub-resources, represented as sub paths below the resource:

* Cluster-scoped subresource: `GET /apis/<GROUP>/<VERSION>/<RESOURCETYPE>/<NAME>/<SUBRESOURCE>`
* Namespace-scoped subresource: `GET /apis/<GROUP>/<VERSION>/namespaces/<NAMESPACE>/<RESOURCETYPE>/<NAME>/<SUBRESOURCE>`

The verbs supported for each subresource will differ depending on the object - see the API documentation more information. It is not possible to access sub-resources across multiple resources - generally a new virtual resource type would be used if that becomes necessary.

## Efficient detection of changes

> To enable clients to build a model of the current state of a cluster, all Kubernetes object resource types are required to support consistent lists and an incremental change notification feed called a **watch**.  Every Kubernetes object has a `resourceVersion` field representing the version of that resource as stored in the underlying database. When retrieving a collection of resources (either namespace or cluster scoped), the response from the server will contain a `resourceVersion` value that can be used to initiate a Watch against the server. The server will return all changes (creates, deletes, and updates) that occur after the supplied `resourceVersion`. This allows a client to fetch the current state and then watch for changes without missing any updates. If the client Watch is disconnected they can restart a new Watch from the last returned `resourceVersion`, or perform a new collection request and begin again. See [Resource Versions](#resource-versions) for more detail.

Reviewer: Could you use code block formatting around the `watch` term instead of double asterisks or capitalization?
Can you cleanup the "will contain" to "contains" and "will return" to "returns"
Edit: 
If a client is disconnected from a `watch`, they can restart a new `watch` from the last returned `resourceVersion`, or perform a new collection request to receive updates again. See [Resource Versions](#resource-versions) for more detail.

Author: I've just realized that there are three meanings of “watch” in Kubernetes - it's a verb, `watch` (eg in RBAC); it's a noun, `Watch`, as well (representing a structure in an API response for example), and then the English word watch describes what you use those jargon terms to achieve.

I guess I need to work out which I think each of these is closest to and adjust styling to match.

Author: Now addressed, I think.

For example:

1. List all of the pods in a given namespace.

   ```
   GET /api/v1/namespaces/test/pods HTTP/1.1

>    …more headers…

Reviewer 3: Please avoid using "…". Use "...".

Author: > Please avoid using … (a single character). Use ... (three dots) when appropriate.

I recommend …; what's the shortcoming there?

Reviewer 3: it is not ASCII character

Author: @tengqm I've no problem with Unicode - we already have 13 localizations and several of those rely on UTF-8 being widely adopted. What's the risk with using characters beyond ASCII's 127?

Reviewer 3: This paragraph is a fenced code block, although it is not code strictly speaking.
In a code block, we rarely see "…", we only see plain dots.

Author: I've omitted this change from the latest revision.

   }
   ```

2. Starting from resource version 10245, receive notifications of any creates, deletes, or updates as a concatenated stream of JSON documents.

   ```
   GET /api/v1/namespaces/test/pods?watch=1&resourceVersion=10245 HTTP/1.1

>    …more headers…

Reviewer: I would not include line 91, `…more headers…`

Author: Here's a thing though. If someone who doesn't know HTTP sees that, will they realize it's only part of a valid request?

Reviewer 3: I'm not aware of any conventions for representing HTTP request and response. I'm not convinced that line 90 will be helpful for readers in this context.

Author: I've omitted this change from the latest revision.

  * `GET /apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE` - return collection of all instances of the resource type in NAMESPACE
  * `GET /apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE/NAME` - return the instance of the resource type with NAME in NAMESPACE

Since a namespace is a cluster-scoped resource type, you can retrieve the list
(“collection”) of all namespaces with `GET /api/v1/namespaces` and details about
a particular namespace with `GET /api/v1/namespaces/NAME`.

* Cluster-scoped subresource: `GET /apis/GROUP/VERSION/RESOURCETYPE/NAME/SUBRESOURCE`
* Namespace-scoped subresource: `GET /apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE/NAME/SUBRESOURCE`

The verbs supported for each subresource will differ depending on the object -
see the [API reference](/docs/reference/kubernetes-api/) for more information. It
is not possible to access sub-resources across multiple resources - generally a new
virtual resource type would be used if that becomes necessary.

## Efficient detection of changes

To enable clients to build a model of the current state of a cluster, all Kubernetes
object resource types are required to support consistent **list** requests and an

> incremental change notification feed called a **watch**.  Every Kubernetes object

Reviewer 2: Watch is not mandatory for aggregated apiservers, and especially for virtual resources. (and in fact is super hard to implement).

Author: Reworded.

has a `resourceVersion` field representing the version of that resource as stored in
the underlying database. When retrieving a collection of resources (either namespace
or cluster scoped), the response from the server will contain a `resourceVersion`
value that can be used to initiate a **watch** against the server.
The server returns all changes (the outcome of **create**s, **delete**s, and
**update**s) that occur after the supplied `resourceVersion`. This allows a client
to fetch the current state and then subscribe to subsequent changes, without missing
any events.

If a client **watch** is disconnected then that client can start a new **watch** from

> the last returned `resourceVersion`; the client could also perform a fresh **get** /

Reviewer 2: optional: could mention bookmarks here. maybe too much detail.

  * `GET /apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE` - return collection of all instances of the resource type in NAMESPACE
  * `GET /apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE/NAME` - return the instance of the resource type with NAME in NAMESPACE

Since a namespace is a cluster-scoped resource type, you can retrieve the list
(“collection”) of all namespaces with `GET /api/v1/namespaces` and details about
a particular namespace with `GET /api/v1/namespaces/NAME`.

* Cluster-scoped subresource: `GET /apis/GROUP/VERSION/RESOURCETYPE/NAME/SUBRESOURCE`
* Namespace-scoped subresource: `GET /apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE/NAME/SUBRESOURCE`

> The verbs supported for each subresource will differ depending on the object -

Reviewer: nit: replace "will differ"
Each sub-resource supports various verbs. For more information, see [API Reference](...).

     "type": "MODIFIED",
     "object": {"kind": "Pod", "apiVersion": "v1", "metadata": {"resourceVersion": "11020", ...}, ...}
   }
   …
   ```

   {{< note >}}
   Although the API server responded with a `Content-Type` header indicating a JSON response,

>    the response to a _watch_ request is in fact a stream of multiple documents. For JSON format, the client can

Reviewer 3: We may want to decide how to write API request verbs, i.e. _watch_ (`_watch_`), or *watch* (`*watch*`), or **watch** (`**watch**`), or `watch` (\`watch\`).
 
Is it okay to say that the response body is a [JSON lines](https://jsonlines.org/) format?

Author: I don't think it is definitely JSONlines, but it might be.

Author: I've gone for `**watch**`.

     "type": "MODIFIED",
     "object": {"kind": "Pod", "apiVersion": "v1", "metadata": {"resourceVersion": "11020", ...}, ...}
   }

>    …

Reviewer: revert this change?

   ```

   {{< note >}}
   Although the API server responded with a `Content-Type` header indicating a JSON response,
   the response to a _watch_ request is in fact a stream of multiple documents. For JSON format, the client can

>    detect the end of each JSON document and begin parting the next one.

Reviewer 3: This extra note is difficult to parse.

Author: Let's find _some_ way to make sure the difference in format is clear.

Reviewer: nit?: s/parting/parsing/
I think you are indicating that the example response represents a partial response.
The `watch` response is composed of multiple JSON documents.

Author: As I understand it, the watch response is invalid JSON that will break some parsers. I think this is relevant to readers, especially anyone implemented a client. Otherwise, a reader might otherwise expect they can use a validating parser and have it just work.

Author: I've omitted this change from the latest revision.

   {{< /note >}}

A given Kubernetes cluster only retains a historical list of changes for a limited time. Clusters using etcd3 preserve changes in the last 5 minutes by default. If a requested Watch operation fails because the historical version of that resource is not available, clients must handle the case by recognizing the HTTP response status 410 (Gone). When a client sees a 410 response it should clear any relevant records from its local cache, perform a new List operation, and request a new Watch that starts from the `resourceVersion` returned by that new List operation.

> 

Reviewer: Questions about formatting the response code as well as list and watch?

When a client receives a `410` response, it must clear any relevant records from its local cache, perform a new `list` operation, and request a new `watch` that starts from the `resourceVersion` returned by that new `list` operation.

Author: I've omitted this change from the latest revision.

> Most Kubernetes client libraries offer some form of standard tool for this Watch retry logic. (In Go this is called a `Reflector` and is located in the `k8s.io/client-go/cache` package.)

Reviewer: nit: substitute "offer some form of" with "provide a standard tool" or "provide a tool to create the `watch` retry logic (In Go this is called a ... )"

Author: I've omitted this change from the latest revision.

### Watch bookmarks

> To mitigate the impact of short history window, Kubernetes provides `bookmark` Watch events. A _bookmark_ is a special kind of event to mark that all changes up to a given `resourceVersion` the client is requesting have already been sent. The object returned in that event is of the type requested by the request, but only `resourceVersion` field is set.

Reviewer: nit: To mitigate the impact of a short history window,

Reviewer: Possible edit?
A _bookmark_ is a special kind of event that marks all changes occurring before a client requests a given `resourceVersion` as sent by the API server. The object returned in the bookmark event is the same as the type set in the client request, but only the `resourceVersion` field is set.

Reviewer 3: IIUC, the bookmark watch event should be spelled as `BOOKMARK`. As shown in the example on line 139.

Author: Addressed, I think.

## Retrieving large results sets in chunks

{{< feature-state for_k8s_version="v1.9" state="beta" >}}

On large clusters, retrieving the collection of some resource types may result in very large responses that can impact the server and client. For instance, a cluster may have tens of thousands of pods, each of which is 1-2KiB of encoded JSON. Retrieving all pods across all namespaces may result in a very large response (10-20MB) and consume a large amount of server resources. The kube-apiserver supports the ability to break a single large collection request into many smaller chunks while preserving the consistency of the total request. Each chunk can be returned sequentially; this chunking reduces the size of each HTTP response message and also allows user-oriented clients to display results incrementally to improve responsiveness. Splitting retrieval over multiple HTTP responses means that if a client needs to retry a fetch, only the missing portion needs to be retransmitted.

> 

Reviewer: What is a "user-oriented" client -- a web browser?
Possible edit?
Each chunk is returned sequentially. Chunking reduces the size of an HTTP response which improves responsiveness in user-oriented clients. When a message is split over multiple HTTP responses and the client retries a fetch, only the missing portion of the message needs to be retransmitted by the API server.

Author: > What is a "user-oriented" client -- a web browser?

Let's save dealing with that to another PR? This one is already quite involved. I'd rather make _some_ improvements now than hold off making any until I fully understand the Kubernetes API.

To retrieve a single list in chunks, two parameters `limit` and `continue` are supported on collection requests. When a List request specifies a `limit`, a field `continue` is returned in the `metadata` field. A client should specify the maximum results they wish to receive in each chunk with `limit`; the API server returns up to `limit` resources in the result. If there are more resources in the collection, the API server response also includes a `continue` field. The client can then pass this `continue` value to the API server in the next request, instructing the API server to return the next chunk of results. By continuing until the server returns an empty `continue` value, the client can consume the full set of results.

> Like a Watch operation, a `continue` token expires after a defined amount of time (5 minutes by default). If more results cannot be returned the API server returns a 410 (Gone) status. On receiving a 410 (Gone) response, the client needs to make a new request (possibly omitting the `limit` parameter, or using a larger `limit`).

Reviewer: Check about formatting response codes and capitalization?

Author: 410 is the right code and isn't capitalized. “Gone” is the text that 410 represents, both in English and as commonly returned as the status string by HTTP/1.x servers. However, this is not an actual HTTP/1.1 response; those look like:
```
HTTP/1.1 410 Gone
…
```

See https://tools.ietf.org/html/rfc7231#section-6.5.9 for an example of idiomatic representation in running text.

For example, if there are 1,253 pods on the cluster and the client wants to receive chunks of 500 pods at a time, the client can request those chunks as follows:

> 1. List all of the Pods on a cluster, retrieving up to 500 Pods each time.

Reviewer: Possible edit?
List all pods in a cluster, retrieving up to 500 pods each time.

Author: In the _API_, you're really dealing with Pods.
Maybe:
```suggestion
1. List all of the `Pods` on a cluster, retrieving up to 500 objects each time.
```

Author: I've omitted this change from the latest revision.

   ```
   GET /api/v1/pods?limit=500 HTTP/1.1
   …more headers…
   ```
   ```
   HTTP/1.1 200 OK
   Content-Type: application/json

   {
     "kind": "PodList",
     "apiVersion": "v1",
     "metadata": {
       "resourceVersion":"10245",
       "continue": "ENCODED_CONTINUE_TOKEN_1",
       ...
     },

>      "items": […] // returns Pods 1-500

Reviewer: Possibly label, `Request` before the request code block
and `Response` before the response code block.
If this is an actual sample response, I would keep the text as presented by the server.

Author: > I would keep the text as presented by the server.

That'll be really long! Can you clarify?

Author: I've omitted this change from the latest revision.

   }
   ```

> 2. Continue the previous call, retrieving the next set of 500 Pods.

Reviewer: Should this be capitalized?

Author: In the API, you're really dealing with Pods.
Maybe:
```suggestion
2. Continue the previous call, retrieving the next set of objects (in this case, 500 `Pods`).
```

   ```
   GET /api/v1/pods?limit=500&continue=ENCODED_CONTINUE_TOKEN_1 HTTP/1.1

>    …more headers…

Reviewer: revert the "more headers"?

Author: I've omitted this change from the latest revision.

       "continue": "ENCODED_CONTINUE_TOKEN_2",
       ...
     },
     "items": […] // returns Pods 501-1000
   }
   ```

3. Continue the previous call, retrieving the remaining 253 pods.

   ```
   GET /api/v1/pods?limit=500&continue=ENCODED_CONTINUE_TOKEN_2 HTTP/1.1
   …more headers…
   ```
   ```
   HTTP/1.1 200 OK
   Content-Type: application/json

   {
     "kind": "PodList",
     "apiVersion": "v1",
     "metadata": {
       "resourceVersion":"10245",
       "continue": "", // continue token is empty because this is the end of the list
       ...
     },
     "items": […] // returns Pods 1001-1253
   }
   ```

> Notice that the `resourceVersion` of the PodList remains constant across each response, indicating thay the control plane providing a consistent snapshot of the cluster's Pods. Pods that are created, updated, or deleted after version `10245` would not be reflected in these responses (the client could however make a new List request at any time). This transaction-like behavior allows clients to break large requests into smaller chunks and then perform a Watch operation on the full set, without missing any updates.

Reviewer: s/thay/that/
Edit:
Notice that the `resourceVersion` of the PodList remains constant across each response, indicating that the control plane provides a consistent snapshot of the cluster's pods.

Author: I've omitted this change from the latest revision.

> ## Alternate representations of resources

Reviewer: Can you add a sentence describing the content in this H2 level section?

Author: I've omitted this change from the latest revision.

### Receiving resources as tables

By default, `kubectl get <kind>` prints a simple tabular representation of one or more instances of a particular resource type.

> It's useful for other API clients to be able to reproduce the same kind of formatted, tabular representation for a resource; this helps make user interfaces consistent.

Reviewer: nit: Is line 232 needed? 
Are you indicating that the client is responsible for formatting the server's response into a table format, not the server OR that: When writing an API client, consider how you want to display the returned resources.

Author: I've omitted this change from the latest revision.

Kubernetes API clients may request the Table representation of objects, delegating specific details of printing to the API server. The Kubernetes API uses HTTP content negotiation based on specialized media types; for example: passing an `Accept` header containing a value of `application/json;as=Table;g=meta.k8s.io;v=v1beta1` with a `GET` call will request that the server return a JSON document that serializes Kubernetes Table objects.

For example, list all of the pods on a cluster, returning a Table serialized as JSON:

```console
GET /api/v1/pods HTTP/1.1
Accept: application/json;as=Table;g=meta.k8s.io;v=v1beta1
…more headers…
```
```
HTTP/1.1 200 OK
Content-Type: application/json

{
    "kind": "Table",
    "apiVersion": "meta.k8s.io/v1beta1",
    …
    "columnDefinitions": [
        …
    ]
}
```

> For API resource types that do not have a custom Table definition known to the cluster, the API server prepares a default Table response that consists of the resource's `name` and `creationTimestamp` fields. For example:

Reviewer: nit: cleanup: For API resource types that do not have a custom `Table` definition, the API server prepares ...

assume bookmarks are returned at any specific interval, nor can clients assume that
the API server will send any `BOOKMARK` event even when requested.

## Retrieving large results sets in chunks

{{< feature-state for_k8s_version="v1.9" state="beta" >}}

On large clusters, retrieving the collection of some resource types may result in
very large responses that can impact the server and client. For instance, a cluster
may have tens of thousands of Pods, each of which is equivalent to roughly 2 KiB of
encoded JSON. Retrieving all pods across all namespaces may result in a very large
response (10-20MB) and consume a large amount of server resources.

Provided that you don't explicitly disable the `APIListChunking` [feature
gate](/docs/reference/command-line-tools-reference/feature-gates/), the Kubernetes
API server supports the ability to break a single large collection request into many
smaller chunks while preserving the consistency of the total request. Each chunk can
be returned sequentially which reduces both the total size of the request and allows
user-oriented clients to display results incrementally to improve responsiveness.

> To retrieve a single list in chunks, two parameters `limit` and `continue` are

Reviewer: nit: two query parameters?

}
```

As a client, you can request `BOOKMARK` events by setting the
`allowWatchBookmarks=true` query parameter to a **watch** request, but you shouldn't
assume bookmarks are returned at any specific interval, nor can clients assume that
the API server will send any `BOOKMARK` event even when requested.

> ## Retrieving large results sets in chunks

Reviewer: nit: Future PR, possibly change this heading.
Chunk large result sets
Pagination

}
```

> Not all API resource types will support a Table response; for example, with a CustomResourceDefinition or an APIService that extends the core Kubernetes API. Clients that must work against all resource types, should specify multiple content types in their `Accept` header to support fallback to non-Tabular JSON:

Reviewer: nit, possible edit:
Not all API resource types support a Table response such as the CustomResourceDefinition or APIService resources. Clients that must work against all resource types should specify multiple content types in their `Accept` header to support fallback to non-Tabular JSON:

Reviewer: Could you repeat this sentence (from below):
Not all API resource types support [Protobuf], specifically those defined via CustomResourceDefinitions or those that are API extensions.

Author: Addressed, I think.

```console
Accept: application/json;as=Table;g=meta.k8s.io;v=v1, application/json
```

## Alternate representations of resources

By default, Kubernetes returns objects serialized to JSON with content type `application/json`. This is the default serialization format for the API. However, clients may request the more efficient Protobuf representation of these objects for better performance at scale. The Kubernetes API implements standard HTTP content type negotiation: passing an `Accept` header with a `GET` call will request that the server return objects in the provided content type, while sending an object in Protobuf to the server for a `PUT` or `POST` call takes the `Content-Type` header. The server will return a `Content-Type` header if the requested format is supported, or the `406 Not acceptable` error if an invalid content type is provided.

> See the [API reference](/docs/reference/kubernetes-api/) for a list of supported content types for each API.

Reviewer: nit: See the Kubernetes API reference ...

included in this response. If the **list** request contained label or field
{{< glossary_tooltip text="selectors" term_id="selector">}} then the number of
remaining items is unknown and the API server does not include a `remainingItemCount`
field in its response.
If the **list** is complete (either because it is not chunking, or because this is the
last chunk), then there are no more remaining items and the API server does not include a
`remainingItemCount` field in its response. The intended use of the `remainingItemCount`
is estimating the size of a collection.

## Collections

In Kubernetes terminology, the response you get from a **list** is
a _collection_. However, Kubernetes defines concrete kinds for
collections of different types of resource. Collections have a kind
named for the resource kind, with `List` appended.

When you query the API for a particular type, all items returned by that query are
of that type.
For example, when you **list** Services, the collection response
has `kind` set to
[`ServiceList`](/docs/reference/kubernetes-api/service-resources/service-v1/#ServiceList); each item in that collection represents a single Service. For example:

> ```

Reviewer: nit: A future PR could split the request from the response

...
```

There are dozens of collection types (such as `PodList`, `ServiceList`,

> and `NodeList`) defined in the Kubernetes API.

Reviewer: Good

resource. Instead, the API server sets its `.metadata.deletionTimestamp` to the current
time.

Once the `.metadata.deletionTimestamp` is set,
{{< glossary_tooltip term_id="controller" text="controllers">}} that act on the object's
finalizers may start performing their cleanup. The controllers can perform finalization
at any time and in any order.

{{< note >}}
Enforced ordering would introduce significant risk of finalization deadlock.

`.metadata.finalizers` is a shared field; any actor with permission to write that
field can reorder it. If the finalizer list is processed in order, this can lead to
a situation in which the component responsible for the first finalizer in the list is
waiting for a signal (field value, external system, or other) produced by a
component responsible for a finalizer later in the list, and the finalization is
unable to proceed.

Without enforced ordering, finalizers are free to order amongst themselves and
are not vulnerable to ordering changes in the list.
{{< /note >}}

After a finalizer completes its tasks, the relevant controller marks that completion
by removing its own entry from the list of finalizers on that resource.

> Once the last finalizer is removed, the resource is actually removed from etcd and,

Reviewer: nit: remove "actually"

Author: I've omitted this change from the latest revision.

consequently, from the API.

See [Garbage Collection](/docs/concepts/workloads/controllers/garbage-collection/) for
details about built-in finalization for cleaning up dependents.

> ## Single- and multiple-resource API verbs {#single-resource-api}

Reviewer: I suggest to remove the dashes in the heading

Author: I've omitted this change from the latest revision.

In the API, verbs Get, Create, Update, Patch, Delete and Proxy support single
resources only.
These verbs with single resource support have no support for submitting
multiple resources together in an ordered or unordered list or transaction.
Clients, including `kubectl`, parse a list of resources and make
a series of single-resource API requests if needed.

The API verbs List and Watch support getting multiple resources;
DeleteCollection supports deleting multiple resources.

## Dry run mode {#dry-run}

 {{< feature-state for_k8s_version="v1.18" state="stable" >}}

> In the Kubernetes API, HTTP verbs that modify data (`POST`, `PUT`, `PATCH`, and `DELETE`) can accept requests in a _dry run_ mode. Dry run mode helps to evaluate a request through the typical request stages (admission chain, validation, merge conflicts) up until persisting objects to storage. The response body for the request is as close as possible to a non-dry-run response. The system guarantees that dry-run requests will not be persisted in storage or have any other side effects.

Reviewer: nit? "non-dry-run" ?
The query parameter is "dryRun".

Author: non-dry-run is idiomatic UK English; I can't guess at the US version - can you recommend a replacement?

Reviewer: The double hyphen word combination looks odd to me.
"dry run" vs "dry-run"

Author: I've omitted this change from the latest revision.

Authorization for dry-run and non-dry-run requests is identical. Thus, to make
a dry-run request, the user must be authorized to make the non-dry-run request.

> For example, to run a dry-run Patch for Deployments, you must have the

Reviewer: nit: "dry run" or "dry-run"
nit: Leave PATCH as `PATCH`

Author: I think `patch` is most appropriate here (matching the RBAC rule) but would need tech review / advice to confirm.

Author: Addressed.

See [Authorization Overview](/docs/reference/access-authn-authz/authorization/).

> ### Generated values for dry-run requests {#generated-values}

Reviewer: Why is dry run hyphenated?

Author: That's how we write it in England.

## Server Side Apply

Kubernetes' [Server Side Apply](/docs/reference/using-api/server-side-apply/)

> feature allows the control plane to tracks managed fields for newly created objects.

Reviewer: ```suggestion
feature allows the control plane to track managed fields for newly created objects.
```

### `ResourceVersion` in metadata {#resourceversion-in-metadata}

Clients find resource versions in resources, including the resources in watch events, and list responses returned from the server:

[v1.meta/ObjectMeta](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#objectmeta-v1-meta) - The `metadata.resourceVersion` of a resource instance identifies the resource version the instance was last modified at.

[v1.meta/ListMeta](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#listmeta-v1-meta) - The `metadata.resourceVersion` of a resource collection (i.e. a list response) identifies the resource version at which the list response was constructed.

> ### `ResourceVersion` parameter {#the-resourceversion-parameter}

Reviewer: Should the `resourceVersion` parameter in a heading list as `ResourceVersion` parameter?

Author: Addressed, I think.

The **get**, **list**, and **watch** operations support the `resourceVersion` parameter.

The exact meaning of this parameter differs depending on the operation and the value of `resourceVersion`.

For get and list, the semantics of resource version are:

**get:**

| resourceVersion unset | resourceVersion="0" | resourceVersion="{value other than 0}" |
|-----------------------|---------------------|----------------------------------------|
| Most Recent           | Any                 | Not older than                         |

**list:**

From version v1.19, Kubernetes API servers support the `resourceVersionMatch` parameter,

> which determines how resourceVersion is applied to list calls.  It is highly

Reviewer: nit: ... how the `resourceVersion` parameter is applied to list calls?

Author: Addressed, I think.

feature, see the section on
[Server Side Apply](/docs/reference/using-api/server-side-apply/).

## Resource versions

Resource versions are strings that identify the server's internal version of an object. Resource versions can be used by clients to determine when objects have changed, or to express data consistency requirements when getting, listing and watching resources. Resource versions must be treated as opaque by clients and passed unmodified back to the server.

{{< note >}}
Clients must not assume resource versions are numeric, and may only compare two resource version for equality (clients must **not** compare resource versions for greater-than or less-than relationships).
{{< /note >}}

### resourceVersion field

Objects and virtual resources have a `resourceVersion` field.

Objects have a [v1.meta/ObjectMeta](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#objectmeta-v1-meta). The `metadata.resourceVersion` of a resource instance identifies the resource version the instance was last modified at.

> Collection responses have a [v1.meta/ListMeta](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#listmeta-v1-meta). The `metadata.resourceVersion` of a resource collection (such as the response to a List) identifies the resource version at which the collection response was constructed.

Reviewer: nit: Collection responses have a [v1.meta/ListMeta](...) object.

Author: How about:
```suggestion
Collection responses have a [v1.meta/ListMeta](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#listmeta-v1-meta) field. The `metadata.resourceVersion` of a resource collection (such as the response to a List) identifies the resource version at which the collection response was constructed.
```
?

This field is present for List responses and for resources in Watch events.

### resourceVersion request parameter

The Get, List and Watch operations support the `resourceVersion` parameter; for example, in the query string.

The exact meaning of this parameter differs depending on the operation and the value of `resourceVersion`.

> For Get and List, the semantics of resource version are:

Reviewer: Check the capitalization?

**List:**

v1.19+ API servers also support the `resourceVersionMatch` parameter, which

> determines how resourceVersion is applied to List calls.  It is highly

Reviewer: nit: `resourceVersion`? OR "how the version is applied to `list` requests"?
nit: It is recommended that ...

  }
```

{{< note >}}
Clients that receive a response in `application/vnd.kubernetes.protobuf` that does
not match the expected prefix should reject the response, as future versions may need
to alter the serialization format in an incompatible way and will do so by changing
the prefix.
{{< /note >}}

## Resource deletion

When you **delete** a resource this takes place in two phases.

> 1. _finalization_

Reviewer: Are the phases defined in this page?

Author: Can we save that for a follow-up? I don't think the original text describes these either.

Reviewer: 👍

Reviewer: 👍

  "Added" events for all resources instances that exist at the starting resource version.
  All following Watch events are for all changes that occurred after the resource version the Watch
  started at.
- **Start at Exact:**  
  Start a Watch at an exact resource version. The watch events are for all changes after the
  provided resource version. Unlike "Get State and Start at Most Recent" and
  "Get State and Start at Any", the Watch is not started with synthetic "Added" events for the
  provided resource version. The client is assumed to already have the initial state at the
  starting resource version since the client provided the resource version.

### "410 Gone" responses

Servers are not required to serve all older resource versions and may return a HTTP 410 (Gone) response if a client requests a `resourceVersion` older than the API server has retained. Clients must be able to tolerate HTTP 410 responses. See [Efficient detection of changes](#efficient-detection-of-changes) for details on how to handle `410 Gone` responses when watching resources.

### Unavailable resource versions

> Servers are not required to serve unrecognized resource versions. List and Get requests for unrecognized resource versions may wait briefly for the resource version to become available; after waiting the server should time out, returning a 504 (Gateway Timeout) response. The 504 response may include a `Retry-After` header than indicates how many seconds a client should wait before retrying the request. The kube-apiserver also identifies these responses with a "Too large resource version" message. Watch requests for a unrecognized resource version may wait indefinitely (until the request times out) for the resource version to become available.

Author: oops, a typo:
```suggestion
Servers are not required to serve unrecognized resource versions. List and Get requests for unrecognized resource versions may wait briefly for the resource version to become available; after waiting the server should time out, returning a 504 (Gateway Timeout) response. The 504 response may include a `Retry-After` header that indicates how many seconds a client should wait before retrying the request. The kube-apiserver also identifies these responses with a "Too large resource version" message. Watch requests for a unrecognized resource version may wait indefinitely (until the request times out) for the resource version to become available.
```

}
```

> When a client first **delete**s a resource, the `.metadata.deletionTimestamp` is set to the current time.

Reviewer: Q:  The original sentence used the verb "deletes" in a general way. When the verb is formatted, does this imply that the client is making a request in a special way?

Author: The client is telling the API server to perform deletion.

Reviewer: OK, yes. I looked at the preview. The partial formatting of the verb does not look as nice. It's okay, but may look better if you format the entire string.

Once the `.metadata.deletionTimestamp` is set, external controllers that act on finalizers
may start performing their cleanup work at any time, in any order.

Order is **not** enforced between finalizers because it would introduce significant
risk of stuck `.metadata.finalizers`.

The `.metadata.finalizers` field is shared: any actor with permission can reorder it.
If the finalizer list were processed in order, then this might lead to a situation
in which the component responsible for the first finalizer in the list is
waiting for some signal (field value, external system, or other) produced by a
component responsible for a finalizer later in the list, resulting in a deadlock.

Without enforced ordering, finalizers are free to order amongst themselves and are
not vulnerable to ordering changes in the list.

Once the last finalizer is removed, the resource is actually removed from etcd.

## Single resource API

> The Kubernetes API verbs **get**, **create**, **update**, **patch**, **delete** and

Reviewer: The original text used the HTTP verbs. Is this change correct?

Author: There is no CREATE verb in HTTP; the original text was perhaps misleading.

Reviewer: Right, you would use `POST`.

Reviewer: nit: This section seems out of place, mostly because the previous sections describe collections.
Possibly move after, https://deploy-preview-25299--kubernetes-io-main-staging.netlify.app/docs/reference/using-api/api-concepts/#resource-uris

Reviewer 2: If listing the k8s verbs, this should include APPLY

**proxy** support single resources only. These verbs with single resource support
have no support for submitting multiple resources together in an ordered or unordered
list or transaction.

When clients (including kubectl) act on a list of resources, the client makes a series
of single-resource API requests, then aggregates the responses if needed.

By contrast, the Kubernetes API verbs **list** and **watch** allow getting multiple
resources, and **deletecollection** allows deleting multiple resources.

## Dry-run

 {{< feature-state for_k8s_version="v1.18" state="stable" >}}

> When you use an HTTP verbs that can modify resources (`POST`, `PUT`, `PATCH`, and

Reviewer: nit: When you use HTTP verbs ...

| _unset_            | limit=\<n\>, _continue unset_     | Most Recent           | Any                                       | Exact                                  |
| _unset_            | limit=\<n\>, continue=\<token\> | Continue Token, Exact | Invalid, treated as Continue Token, Exact | Invalid, HTTP `400 Bad Request`        |
| `resourceVersionMatch=Exact`        | _limit unset_                 | Invalid               | Invalid                                   | Exact                                  |
| `resourceVersionMatch=Exact`        | limit=\<n\>, _continue unset_ | Invalid               | Invalid                                   | Exact                                  |
| `resourceVersionMatch=NotOlderThan` | _limit unset_                 | Invalid               | Any                                       | Not older than                         |
| `resourceVersionMatch=NotOlderThan` | limit=\<n\>, _continue unset_ | Invalid               | Any                                       | Not older than                         |

{{< /table >}}

{{< note >}}

> If your clusterserver does not honor the `resourceVersionMatch` parameter, the behavior is the same as if you did not set it.

Reviewer: nit: Not sure what this is: clusterserver -> cluster's API server?

Author: Yep, that's right. Fixed.

By contrast, the Kubernetes API verbs **list** and **watch** allow getting multiple
resources, and **deletecollection** allows deleting multiple resources.

## Dry-run

 {{< feature-state for_k8s_version="v1.18" state="stable" >}}

When you use an HTTP verbs that can modify resources (`POST`, `PUT`, `PATCH`, and
`DELETE`), you can submit your request in a _dry run_ mode. Dry run mode helps to
evaluate a request through the typical request stages (admission chain, validation,
merge conflicts) up until persisting objects to storage. The response body for the
request is as close as possible to a non-dry-run response. Kubernetes guarantees
that dry-run requests will not be persisted in storage or have any other side effects.

### Make a dry-run request

Dry-run is triggered by setting the `dryRun` query parameter. This parameter is a
string, working as an enum, and the only accepted values are:

[empty]
: Allow side effects. You request this with a query string such as `?dryRun`
  or `?dryRun&pretty=true`. The response is the final object that would have been
  persisted, or an error if the request could not be fulfilled.

> `All`

Reviewer: Check the preview. The text in the description list is rendering differently. The first element uses one type of font compared to the second element (formatted string "All").

Author: `All` is a literal value. I changed [empty] to [no value set]; is that more clear?

  **list** calls after the initial paginated **list**.

{{< note >}}
When you **list** resources and receive a collection response, the response includes the
[metadata](/docs/reference/generated/kubernetes-api/v1.21/#listmeta-v1-meta) of the collection as
well as [object metadata](/docs/reference/generated/kubernetes-api/v1.21/#listmeta-v1-meta)
for each item in that collection. For individual objects found within a collection response,
`.metadata.resourceVersion` tracks when that object was last updated, and not how up-to-date
the object is when served.
{{< /note >}}

When using `resourceVersionMatch=NotOlderThan` and limit is set, clients must
handle HTTP 410 "Gone" responses. For example, the client might retry with a
newer `resourceVersion` or fall back to `resourceVersion=""`.

When using `resourceVersionMatch=Exact` and `limit` is unset, clients must
verify that the collection's `.metadata.resourceVersion` matches
the requested `resourceVersion`, and handle the case where it does not. For
example, the client might fall back to a request with `limit` set.

### Semantics for **watch**

For **watch**, the semantics of resource version are:

> 

Reviewer: Check lines 892 - 894, and then lines started at line 904, The meaning of those **watch** semantics are:
I think there may be some duplication,
https://deploy-preview-25299--kubernetes-io-main-staging.netlify.app/docs/reference/using-api/api-concepts/#semantics-for-watch

Author: It's not very clear; maybe a follow-up PR could try to make this easier to read?
